export_tsvs.py

```diff
@@ -1,64 +1,138 @@
+		"-f", "--overwrite-checkpoints",
+		action="store_true",
+		help="Overwrite/update all hail checkpoints.",
+	)
+	p.add_argument(
+		"sample_ids_path",
+		help="A text file containing one sample id per line",
+	)
+	args = p.parse_args()
+
+	return args
+
+
+def read_sample_ids(sample_ids_path, n_sample_ids_to_print = 10):
+	"""Read sample ids file.
+
+	Args:
+		sample_ids_path (str): sample ids path
+		n_sample_ids_to_print (int): log no more than this many sample ids to stdout.
+	Return:
+		list: sample id strings
+	"""
+	sample_ids = []
+	with hl.hadoop_open(sample_ids_path) if sample_ids_path.startswith("gs://" ) else open(sample_ids_path, "rt") as f:
+		for i, line in enumerate(f):
+			sample_id = line.rstrip("\n")
+			sample_ids.append(sample_id)
+
+			if i <= n_sample_ids_to_print:
+				logging.info(sample_id)
+				if i == n_sample_ids_to_print and n_sample_ids_to_print > 0:
+					logging.info("...")
+
+	logging.info(f"Parsed {len(sample_ids)} sample ids from {sample_ids_path}")
+
+	return sample_ids
+
+
+def explode_table_by_sample(ht):
+	"""Explode variant-level ht by sample."""
+
+	logging.info("Input schema:")
+	ht.describe()
+
+	ht = ht.annotate(
+		samples_w_het_var=ht.samples_w_het_var.map(lambda x: hl.struct(S=x.S, GQ=x.GQ, het_or_hom_or_hemi=1)),
+		samples_w_hom_var=ht.samples_w_hom_var.map(lambda x: hl.struct(S=x.S, GQ=x.GQ, het_or_hom_or_hemi=2)),
+		samples_w_hemi_var=ht.samples_w_hemi_var.map(lambda x: hl.struct(S=x.S, GQ=x.GQ, het_or_hom_or_hemi=3)),
+	)
+
+	ht = ht.select(samples=ht.samples_w_het_var.extend(ht.samples_w_hom_var.extend(ht.samples_w_hemi_var)))
+
+	ht = ht.explode(ht.samples)
+	ht = ht.key_by(ht.locus)  # keep locus key to work around bug in naive_coalesce(..) when key is empty (https://github.com/hail-is/hail/issues/8138)
+	ht = ht.transmute(
+		#chrom=ht.locus.contig.replace("chr", ""),
+		#pos=ht.locus.position,
+		ref=ht.alleles[0],
+		alt=ht.alleles[1],
+		het_or_hom_or_hemi=ht.samples.het_or_hom_or_hemi,
+		GQ=ht.samples.GQ,
+		S=ht.samples.S,
+	)
+
+	return ht
+
+
+def export_per_sample_tsvs(ht, sample_ids, output_bucket_path, n_partitions_per_sample):
+	"""Iterate over sample_ids and export all records with that sample id to separate tsv(s).
+
+	Args:
+		ht (hail table): output of explode_table_by_sample(..)
+		sample_ids (list): collection of sample id strings
+		output_bucket_path (str):
+		n_partitions_per_sample (int):
+	"""
+
+	logging.info("Output schema:")
+	ht.describe()
+
+	for s in sorted(sample_ids):
+		per_sample_ht = ht.filter(ht.S==s, keep=True)
+		per_sample_ht = per_sample_ht.drop(per_sample_ht.S)  # drop redundant column. Sample id is in the .tsv filename.
+		if n_partitions_per_sample > 1:
+			tsv_output_path = os.path.join(output_bucket_path, s)
+			per_sample_ht = per_sample_ht.naive_coalesce(n_partitions_per_sample)
+			per_sample_ht.export(tsv_output_path, parallel="separate_header")
+		else:
+			tsv_output_path = os.path.join(output_bucket_path, f"{s}.tsv.bgz")
```

Comment:
I think if the input ht has multiple partitions you will need to still explicitly specify `per_sample_ht = per_sample_ht.naive_coalesce(1)` -- otherwise after filtering to the sample, Hail will maintain the original number of partitions from the input ht (unless things have changed recently)

Reply:
In this `else` it writes out a single .tsv file, so I don't see where the naive_coalesce would help.